File: cdm/cpp/scenario/SEConditionManager.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "scenario/SEConditionManager.h"
//Patient Conditions
#include "patient/conditions/SEChronicAnemia.h"
#include "patient/conditions/SEChronicObstructivePulmonaryDisease.h"
#include "patient/conditions/SEChronicHeartFailure.h"
#include "patient/conditions/SEChronicRenalStenosis.h"
#include "patient/conditions/SEChronicVentricularSystolicDysfunction.h"
#include "patient/conditions/SEConsumeMeal.h"
#include "patient/conditions/SELobarPneumonia.h"
#include "patient/conditions/SEChronicPericardialEffusion.h"
#include "patient/conditions/SEImpairedAlveolarExchange.h"
// Environment Conditions
#include "system/environment/conditions/SEInitialEnvironmentConditions.h"
#include "substance/SESubstance.h"
#include "substance/SESubstanceManager.h"
#include "io/protobuf/PBScenario.h"

SEConditionManager::SEConditionManager(SESubstanceManager& substances) : Loggable(substances.GetLogger()), m_Substances(substances)
{
  m_Anemia = nullptr;
  m_COPD = nullptr;
  m_ChronicVentricularSystolicDysfunction = nullptr;
  m_RenalStenosis = nullptr;
  m_ConsumeMeal = nullptr;
  m_LobarPneumonia = nullptr;
  m_PericardialEffusion = nullptr;
  m_ImpairedAlveolarExchange = nullptr;
  m_InitialEnvironmentConditions = nullptr;
}

SEConditionManager::~SEConditionManager()
{
  Clear();
}

void SEConditionManager::Clear()
{
  SAFE_DELETE(m_Anemia);
  SAFE_DELETE(m_COPD);
  SAFE_DELETE(m_ChronicVentricularSystolicDysfunction);
  SAFE_DELETE(m_RenalStenosis);
  SAFE_DELETE(m_ConsumeMeal);
  SAFE_DELETE(m_LobarPneumonia);
  SAFE_DELETE(m_PericardialEffusion);
  SAFE_DELETE(m_ImpairedAlveolarExchange);
  SAFE_DELETE(m_InitialEnvironmentConditions);
}

bool SEConditionManager::SerializeToString(std::string& output, SerializationMode m) const
{
  return PBScenario::SerializeToString(*this, output, m);
}
bool SEConditionManager::SerializeToFile(const std::string& filename, SerializationMode m) const
{
  return PBScenario::SerializeToFile(*this, filename, m);
}
bool SEConditionManager::SerializeFromString(const std::string& src, SerializationMode m)
{
  return PBScenario::SerializeFromString(src, *this, m);
}
bool SEConditionManager::SerializeFromFile(const std::string& filename, SerializationMode m)
{
  return PBScenario::SerializeFromFile(filename, *this, m);
}

bool SEConditionManager::ProcessCondition(const SECondition& condition)
{
  if (!condition.IsValid())
  {
    m_ss << "Ignoring invalid condition : " << condition;
    Error(m_ss);
    return false;
  }

  if (dynamic_cast<const SEPatientCondition*>(&condition) != nullptr)
  {

    const SEChronicAnemia* a = dynamic_cast<const SEChronicAnemia*>(&condition);
    if (a != nullptr)
    {
      if (HasChronicAnemia())
      {
        Error("Cannot have multiple Anemia conditions");
        return false;
      }
      m_Anemia = new SEChronicAnemia();
      m_Anemia->Copy(*a);
      return true;
    }

    const SEChronicObstructivePulmonaryDisease* copd = dynamic_cast<const SEChronicObstructivePulmonaryDisease*>(&condition);
    if (copd != nullptr)
    {
      if (HasChronicObstructivePulmonaryDisease())
      {
        Error("Cannot have multiple COPD conditions");
        return false;
      }
      m_COPD = new SEChronicObstructivePulmonaryDisease();
      m_COPD->Copy(*copd);
      return true;
    }

    const SEChronicHeartFailure* h = dynamic_cast<const SEChronicHeartFailure*>(&condition);
    if (h != nullptr)
    {
      if (HasChronicHeartFailure())
      {
        Error("Cannot have multiple Heart Failure conditions");
        return false;
      }
      const SEChronicVentricularSystolicDysfunction* vsd = dynamic_cast<const SEChronicVentricularSystolicDysfunction*>(&condition);
      if (vsd != nullptr)
      {
        m_ChronicVentricularSystolicDysfunction = new SEChronicVentricularSystolicDysfunction();
        m_ChronicVentricularSystolicDysfunction->Copy(*vsd);
        return true;
      }
      Error("Unknown Heart Failure condition");
      return false;
    }

    const SEChronicPericardialEffusion* pe = dynamic_cast<const SEChronicPericardialEffusion*>(&condition);
    if (pe != nullptr)
    {
      if (HasChronicPericardialEffusion())
      {
        Error("Cannot have multiple Pericardial Effusion conditions");
        return false;
      }
      m_PericardialEffusion = new SEChronicPericardialEffusion();
      m_PericardialEffusion->Copy(*pe);
      return true;
    }

    const SEChronicRenalStenosis* r = dynamic_cast<const SEChronicRenalStenosis*>(&condition);
    if (r != nullptr)
    {
      if (HasChronicRenalStenosis())
      {
        Error("Cannot have multiple Renal Stenosis conditions");
        return false;
      }
      m_RenalStenosis = new SEChronicRenalStenosis();
      m_RenalStenosis->Copy(*r);
      return true;
    }

    const SEConsumeMeal* g = dynamic_cast<const SEConsumeMeal*>(&condition);
    if (g != nullptr)
    {
      if (HasConsumeMeal())
      {
        Error("Cannot have multiple Gut Nutrient conditions");
        return false;
      }
      m_ConsumeMeal = new SEConsumeMeal();
      m_ConsumeMeal->Copy(*g);
      return true;
    }

    const SEImpairedAlveolarExchange* iae = dynamic_cast<const SEImpairedAlveolarExchange*>(&condition);
    if (iae != nullptr)
    {
      if (HasImpairedAlveolarExchange())
      {
        Error("Cannot have multiple Impaired Alveolar Exchange conditions");
        return false;
      }
      m_ImpairedAlveolarExchange = new SEImpairedAlveolarExchange();
      m_ImpairedAlveolarExchange->Copy(*iae);
      return true;
    }

    const SELobarPneumonia* lp = dynamic_cast<const SELobarPneumonia*>(&condition);
    if (lp != nullptr)
    {
      if (HasLobarPneumonia())
      {
        Error("Cannot have multiple Lobar Pneumonia conditions");
        return false;
      }
      m_LobarPneumonia = new SELobarPneumonia();
      m_LobarPneumonia->Copy(*lp);
      return true;
    }
  }

  if (dynamic_cast<const SEEnvironmentCondition*>(&condition) != nullptr)
  {
    const SEInitialEnvironmentConditions* ie = dynamic_cast<const SEInitialEnvironmentConditions*>(&condition);
    if (ie != nullptr)
    {
      if (HasInitialEnvironmentConditions())
      {
        Error("Cannot have multiple Initial Environment conditions");
        return false;
      }
      m_InitialEnvironmentConditions = new SEInitialEnvironmentConditions(m_Substances);
      m_InitialEnvironmentConditions->Copy(*ie);
      return true;
    }
  }
  
  /// \error Unsupported Condition
  Error("Unsupported Condition");
  return false;
}

bool SEConditionManager::HasChronicAnemia() const
{
  return m_Anemia == nullptr ? false : m_Anemia->IsValid();
}
SEChronicAnemia* SEConditionManager::GetChronicAnemia()
{
  return m_Anemia;
}
const SEChronicAnemia* SEConditionManager::GetChronicAnemia() const
{
  return m_Anemia;
}

bool SEConditionManager::HasChronicObstructivePulmonaryDisease() const
{
  return m_COPD == nullptr ? false : m_COPD->IsValid();
}
SEChronicObstructivePulmonaryDisease* SEConditionManager::GetChronicObstructivePulmonaryDisease()
{
  return m_COPD;
}
const SEChronicObstructivePulmonaryDisease* SEConditionManager::GetChronicObstructivePulmonaryDisease() const
{
  return m_COPD;
}

bool SEConditionManager::HasChronicHeartFailure() const
{
  return m_ChronicVentricularSystolicDysfunction != nullptr;
}
bool SEConditionManager::HasChronicVentricularSystolicDysfunction() const
{
  return m_ChronicVentricularSystolicDysfunction == nullptr ? false : m_ChronicVentricularSystolicDysfunction->IsValid();
}
SEChronicVentricularSystolicDysfunction* SEConditionManager::GetChronicVentricularSystolicDysfunction()
{
  return m_ChronicVentricularSystolicDysfunction;
}
const SEChronicVentricularSystolicDysfunction* SEConditionManager::GetChronicVentricularSystolicDysfunction() const
{
  return m_ChronicVentricularSystolicDysfunction;
}

bool SEConditionManager::HasChronicPericardialEffusion() const
{
  return m_PericardialEffusion == nullptr ? false : m_PericardialEffusion->IsValid();
}
SEChronicPericardialEffusion* SEConditionManager::GetChronicPericardialEffusion()
{
  return m_PericardialEffusion;
}
const SEChronicPericardialEffusion* SEConditionManager::GetChronicPericardialEffusion() const
{
  return m_PericardialEffusion;
}

bool SEConditionManager::HasChronicRenalStenosis() const
{
  return m_RenalStenosis == nullptr ? false : m_RenalStenosis->IsValid();
}
SEChronicRenalStenosis* SEConditionManager::GetChronicRenalStenosis()
{
  return m_RenalStenosis;
}
const SEChronicRenalStenosis* SEConditionManager::GetChronicRenalStenosis() const
{
  return m_RenalStenosis;
}


bool SEConditionManager::HasConsumeMeal() const
{
  return m_ConsumeMeal == nullptr ? false : m_ConsumeMeal->IsValid();
}
SEConsumeMeal* SEConditionManager::GetConsumeMeal()
{
  return m_ConsumeMeal;
}
const SEConsumeMeal* SEConditionManager::GetConsumeMeal() const
{
  return m_ConsumeMeal;
}

bool SEConditionManager::HasImpairedAlveolarExchange() const
{
  return m_ImpairedAlveolarExchange == nullptr ? false : m_ImpairedAlveolarExchange->IsValid();
}
SEImpairedAlveolarExchange* SEConditionManager::GetImpairedAlveolarExchange()
{
  return m_ImpairedAlveolarExchange;
}
const SEImpairedAlveolarExchange* SEConditionManager::GetImpairedAlveolarExchange() const
{
  return m_ImpairedAlveolarExchange;
}

bool SEConditionManager::HasLobarPneumonia() const
{
  return m_LobarPneumonia == nullptr ? false : m_LobarPneumonia->IsValid();
}
SELobarPneumonia* SEConditionManager::GetLobarPneumonia()
{
  return m_LobarPneumonia;
}
const SELobarPneumonia* SEConditionManager::GetLobarPneumonia() const
{
  return m_LobarPneumonia;
}

bool SEConditionManager::HasInitialEnvironmentConditions() const
{
  return m_InitialEnvironmentConditions == nullptr ? false : m_InitialEnvironmentConditions->IsValid();
}
SEInitialEnvironmentConditions* SEConditionManager::GetInitialEnvironmentConditions()
{
  return m_InitialEnvironmentConditions;
}
const SEInitialEnvironmentConditions* SEConditionManager::GetInitialEnvironmentConditions() const
{
  return m_InitialEnvironmentConditions;
}

void SEConditionManager::GetActiveConditions(std::vector<const SECondition*>& conditions) const
{
  if (HasChronicAnemia())
    conditions.push_back(GetChronicAnemia());
  if (HasChronicObstructivePulmonaryDisease())
    conditions.push_back(GetChronicObstructivePulmonaryDisease());
  if (HasChronicVentricularSystolicDysfunction())
    conditions.push_back(GetChronicVentricularSystolicDysfunction());
  if (HasChronicPericardialEffusion())
    conditions.push_back(GetChronicPericardialEffusion());
  if (HasChronicRenalStenosis())
    conditions.push_back(GetChronicRenalStenosis());
  if (HasConsumeMeal())
    conditions.push_back(GetConsumeMeal());
  if (HasImpairedAlveolarExchange())
    conditions.push_back(GetImpairedAlveolarExchange());
  if (HasLobarPneumonia())
    conditions.push_back(GetLobarPneumonia());
  if (HasInitialEnvironmentConditions())
    conditions.push_back(GetInitialEnvironmentConditions());
}

File: cdm/csharp/engine/SEConditionManager.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

public class SEConditionManager
{
  protected SEAcuteRespiratoryDistressSyndrome ards;
  protected SEChronicAnemia anemia;
  //protected SEConsumeMeal consume_meal;
  protected SEChronicObstructivePulmonaryDisease copd;
  protected SEChronicVentricularSystolicDysfunction cvsd;
  protected SEImpairedAlveolarExchange impaired_alveolar_exchange;
  protected SEChronicPericardialEffusion pericardial_effusion;
  protected SELobarPneumonia lobar_pneumonia;
  protected SEPulmonaryFibrosis pulmonary_fibrosis;
  protected SEChronicRenalStenosis renal_stenosis;
  protected SESepsis sepsis;

  //protected SEInitialEnvironmentConditions* initial_environment_conditions;

  public SEConditionManager()
  {
    Clear();
  }

  public void Clear()
  {
    ards = null;
    anemia = null;
    //consume_meal = null;
    copd = null;
    cvsd = null;
    impaired_alveolar_exchange = null;
    pericardial_effusion = null;
    lobar_pneumonia = null;
    pulmonary_fibrosis = null;
    renal_stenosis = null;
    sepsis = null;
    //initial_environment_conditions = null;
  }

  public bool IsEmpty()
  {
    if (HasAcuteRespiratoryDistressSyndrome())
      return false;
    if (HasChronicAnemia())
      return false;
    if (HasChronicObstructivePulmonaryDisease())
      return false;
    if (HasChronicVentricularSystolicDysfunction())
      return false;
    if (HasChronicPericardialEffusion())
      return false;
    if (HasChronicRenalStenosis())
      return false;
    //if (HasConsumeMeal())
    //  return false;
    if (HasImpairedAlveolarExchange())
      return false;
    if (HasLobarPneumonia())
      return false;
    if (HasPulmonaryFibrosis())
      return false;
    if (HasSepsis())
      return false;

    //if (HasInitialEnvironmentConditions())
    //  return false;
    return true;
  }

  public bool Add(SECondition c)
  {
    if (c.GetType().IsAssignableFrom(typeof(SEAcuteRespiratoryDistressSyndrome)))
    {
      ards = (SEAcuteRespiratoryDistressSyndrome)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SEChronicAnemia)))
    {
      anemia = (SEChronicAnemia)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SEChronicObstructivePulmonaryDisease)))
    {
      copd = (SEChronicObstructivePulmonaryDisease)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SEChronicVentricularSystolicDysfunction)))
    {
      cvsd = (SEChronicVentricularSystolicDysfunction)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SEChronicPericardialEffusion)))
    {
      pericardial_effusion = (SEChronicPericardialEffusion)c;
      return true;
    }
    //if (c.GetType().IsAssignableFrom(typeof(SEConsumeMeal)))
    //{
    //  consume_meal = (SEConsumeMeal)c;
    //  return true;
    //}
    if (c.GetType().IsAssignableFrom(typeof(SEChronicRenalStenosis)))
    {
      renal_stenosis = (SEChronicRenalStenosis)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SEImpairedAlveolarExchange)))
    {
      impaired_alveolar_exchange = (SEImpairedAlveolarExchange)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SELobarPneumonia)))
    {
      lobar_pneumonia = (SELobarPneumonia)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SEPulmonaryFibrosis)))
    {
      pulmonary_fibrosis = (SEPulmonaryFibrosis)c;
      return true;
    }
    if (c.GetType().IsAssignableFrom(typeof(SESepsis)))
    {
      sepsis = (SESepsis)c;
      return true;
    }

    //if (c.GetType().IsAssignableFrom(typeof(SEInitialEnvironmentalConditions)))
    //{
    //  initial_environmental_conditions = (SEInitialEnvironmentalConditions)c;
    //  return true;
    //}
    return false;
  }

  public bool HasAcuteRespiratoryDistressSyndrome()
  {
    return ards == null ? false : ards.IsValid();
  }
  public SEAcuteRespiratoryDistressSyndrome GetAcuteRespiratoryDistressSyndrome()
  {
    if (ards == null)
      ards = new SEAcuteRespiratoryDistressSyndrome();
    return ards;
  }
  public void RemoveAcuteRespiratoryDistressSyndrome()
  {
    ards = null;
  }

  public bool HasChronicAnemia()
  {
    return anemia == null ? false : anemia.IsValid();
  }
  public SEChronicAnemia GetChronicAnemia()
  {
    if (anemia == null)
      anemia = new SEChronicAnemia();
    return anemia;
  }
  public void RemoveChronicAnemia()
  {
    anemia = null;
  }

  public bool HasChronicObstructivePulmonaryDisease()
  {
    return copd == null ? false : copd.IsValid();
  }
  public SEChronicObstructivePulmonaryDisease GetChronicObstructivePulmonaryDisease()
  {
    if (copd == null)
      copd = new SEChronicObstructivePulmonaryDisease();
    return copd;
  }
  public void RemoveChronicObstructivePulmonaryDisease()
  {
    copd = null;
  }

  public bool HasChronicHeartFailure()
  {
    return HasChronicVentricularSystolicDysfunction();
  }
  public bool HasChronicVentricularSystolicDysfunction()
  {
    return cvsd == null ? false : cvsd.IsValid();
  }
  public SEChronicVentricularSystolicDysfunction GetChronicVentricularSystolicDysfunction()
  {
    if (cvsd == null)
      cvsd = new SEChronicVentricularSystolicDysfunction();
    return cvsd;
  }
  public void RemoveChronicVentricularSystolicDysfunction()
  {
    cvsd = null;
  }

  public bool HasChronicPericardialEffusion()
  {
    return pericardial_effusion == null ? false : pericardial_effusion.IsValid();
  }
  public SEChronicPericardialEffusion GetChronicPericardialEffusion()
  {
    if (pericardial_effusion == null)
      pericardial_effusion = new SEChronicPericardialEffusion();
    return pericardial_effusion;
  }
  public void RemoveChronicPericardialEffusion()
  {
    pericardial_effusion = null;
  }

  public bool HasChronicRenalStenosis()
  {
    return renal_stenosis == null ? false : renal_stenosis.IsValid();
  }
  public SEChronicRenalStenosis GetChronicRenalStenosis()
  {
    if (renal_stenosis == null)
      renal_stenosis = new SEChronicRenalStenosis();
    return renal_stenosis;
  }
  public void RemoveChronicRenalStenosis()
  {
    renal_stenosis = null;
  }

  //public bool HasConsumeMeal()
  //public SEConsumeMeal GetConsumeMeal()
  //public void RemoveConsumeMeal()

  public bool HasImpairedAlveolarExchange()
  {
    return impaired_alveolar_exchange == null ? false : impaired_alveolar_exchange.IsValid();
  }
  public SEImpairedAlveolarExchange GetImpairedAlveolarExchange()
  {
    if (impaired_alveolar_exchange == null)
      impaired_alveolar_exchange = new SEImpairedAlveolarExchange();
    return impaired_alveolar_exchange;
  }
  public void RemoveImpairedAlveolarExchange()
  {
    impaired_alveolar_exchange = null;
  }

  public bool HasLobarPneumonia()
  {
    return lobar_pneumonia == null ? false : lobar_pneumonia.IsValid();
  }
  public SELobarPneumonia GetLobarPneumonia()
  {
    if (lobar_pneumonia == null)
      lobar_pneumonia = new SELobarPneumonia();
    return lobar_pneumonia;
  }
  public void RemoveLobarPneumonia()
  {
    lobar_pneumonia = null;
  }

  public bool HasPulmonaryFibrosis()
  {
    return pulmonary_fibrosis == null ? false : pulmonary_fibrosis.IsValid();
  }
  public SEPulmonaryFibrosis GetPulmonaryFibrosis()
  {
    if (pulmonary_fibrosis == null)
      pulmonary_fibrosis = new SEPulmonaryFibrosis();
    return pulmonary_fibrosis;
  }
  public void RemovePulmonaryFibrosis()
  {
    pulmonary_fibrosis = null;
  }

  public bool HasSepsis()
  {
    return sepsis == null ? false : sepsis.IsValid();
  }
  public SESepsis GetSepsis()
  {
    if (sepsis == null)
      sepsis = new SESepsis();
    return sepsis;
  }
  public void RemoveSepsis()
  {
    sepsis = null;
  }

  // Environment Conditions
  //public bool HasInitialEnvironmentalConditions()
  //public SEInitialEnvironmentConditions GetInitialEnvironmentalConditions()
  //public void RemoveInitialEnvironmentalConditions()
}
